src/stan/io/json/json_data_handler.hpp

```diff
@@ -7,174 +7,527 @@
+      key = boost::algorithm::join(key_stack, ".");
+      unexpected_error(key);
+    }
+    slot_dims_map[key] = update;
+  }
+
+  void promote_to_double() {
+    if (int_slots_map[key_str()]) {
+      int_slots_map[key_str()] = false;
+      values_r.reserve(values_i.size());
+      values_r.insert(values_r.end(), values_i.begin(), values_i.end());
+      array_start_r = array_start_i;
+      values_i.clear();
+      array_start_i = 0;
+    }
+  }
+
+  /* Save non-tuple vars and innermost tuple slots to vars_i and vars_r.
+   * For arrays of tuples we need to check that new elements are consistent
+   * with previous tuple elements.
+   */
+  void save_key_value_pair() {
+    if (key_stack.empty())
+      return;
+    std::string key = key_str();
+    if (slot_types_map.count(key) < 1)
+      unexpected_error(key);
+    if (slot_types_map[key] == meta_type::SCALAR
+        || slot_types_map[key] == meta_type::ARRAY) {
+      bool is_new = (vars_r.count(key) == 0 && vars_i.count(key) == 0);
+      bool is_int = int_slots_map[key];
+      bool is_real = vars_r.count(key) == 1;
+      bool was_int = vars_i.count(key) == 1;
+      std::vector<size_t> dims;
+      if (slot_dims_map.count(key) == 1)
+        dims = slot_dims_map[key].dims;
+      if (is_new) {
+        var_types_map[key] = slot_types_map[key];
+        if (is_int) {
+          std::pair<std::vector<int>, std::vector<size_t>> pair;
+          pair = make_pair(values_i, dims);
+          vars_i[key] = pair;
+        } else {
+          std::pair<std::vector<double>, std::vector<size_t>> pair;
+          pair = make_pair(values_r, dims);
+          vars_r[key] = pair;
+        }
+      } else {
+        if (slot_types_map[key_stack[0]] != meta_type::ARRAY_OF_TUPLES)
+          unexpected_error(key);
+        var_types_map[key] = meta_type::ARRAY;
+        std::vector<size_t> dims = slot_dims_map[key].dims;
+        if ((!is_int && was_int) || (is_int && is_real)) {  // promote to double
+          std::vector<double> values_tmp;
+          for (auto& x : vars_i[key].first) {
+            values_tmp.push_back(x);
+          }
+          for (auto& x : values_r)
+            values_tmp.push_back(x);
+          std::pair<std::vector<double>, std::vector<size_t>> pair;
+          pair = make_pair(values_tmp, dims);
+          vars_r[key] = pair;
+          vars_i.erase(key);
+        } else if (is_int) {
+          for (auto& x : values_i)
+            vars_i[key].first.push_back(x);
+          vars_i[key].second = dims;
+        } else {
+          for (auto& x : values_r)
+            vars_r[key].first.push_back(x);
+          vars_r[key].second = dims;
+        }
+      }
+    }
+    key_stack.pop_back();
+  }
+
+  /* Process array variables
+   *  a. for array of tuples, concatenate dimensions
+   *  b. if multi-dim array, convert vector of values
+   *      from row-major order to column-major order.
+   * Update vars_i and vars_r accordingly.
+   */
+  void convert_arrays() {
```

Comment:
I think this function is doing the wrong thing for arrays inside tuples.

Given the decl ` array[2] tuple(real, simplex[4]) t;` and json

```json
{
  "t": [
    {
      "1": 5.8215377737734162,
      "2": [
        0.014580671457001414, 0.30736191423550285, 0.2920816565582352,
        0.38597575774926052
      ]
    },
    {
      "1": 5.6235875354192935,
      "2": [
        0.26041728896763167, 0.29828428322619338, 0.22664074293149306,
        0.2146576848746819
      ]
    }
  ]
}
```

I would expect that `vals_r("t.2")` would return `{0.014580671457001414, 0.30736191423550285, 0.2920816565582352, 0.38597575774926052, 0.26041728896763167, 0.29828428322619338, 0.22664074293149306, 0.2146576848746819}`

E.g. the returned vector is the result of concatinating `t[0].2` and `t[1].2`.

The actual result currently returned is `{0.0145807, 0.260417, 0.307362, 0.298284, 0.292082, 0.226641, 0.385976, 0.214658}`

Which is the result of striping `t[0].2` and `t[1].2`

The later would be difficult for the compiler to correctly un-fold, and I don't think it makes sense to "colum-majorize" these tuple elements

Reply:
This wasn't explicit in the design doc, I've updated the example there to clarify

Reply:
given that we've got the `slot_types_map`, we should be able to handle this.

but when would someone want to do this in a Stan program?  this means that a 2-D array is sometimes row-major, sometimes column-major.   do you have examples of this kinds of operations a user would want to do and why?  is it because we expect users to iterate over an array of tuples element by element?

Reply:
>  this means that a 2-D array is sometimes row-major, sometimes column-major.

Any 2-D array declared by the user should still be column-major. But, if you have an array of tuples, and an element of those tuples is an array, it creates a "phantom" 2-D array in the `var_context` which needs to be read into tuples chunk-by-chunk.

In pseudocode, here is how the `t` above is read into memory

```
vector<tuple<double, Eigen::Vector>> t = //init to NANs;

vector<double> t_1_flat = context.vars_r("t.1");
int t_2_flat_pos = 1;
vector<double> t_2_flat = context.vars_r("t.2")
int t_2_flat_pos = 1;

for ( i = 0; i < 2 < i++) {
  tuple(double, Eigen::Vector> temp;
  std::get<0>(temp) = t_1_flat[t_1_flat_pos]; 
  t_2_flat_pos = t_2_flat_pos + 1;

  std::get<1>(temp) = t_2_flat[t_2_flat_pos: t_2_flat_pos + 4]; // IMPORTANT LINE HERE
  
  t[i] = temp;
}
```

If we treat "t.2" like any other 2D array, the "`IMPORTANT LINE HERE`" part goes from a simple indexing slice to a very expensive memory striping problem. 

You make a good point about what happens if you have a `array[2] tuple(double, array[2,3]) t2;`, where the inner-array _is_ actually 2D, so now the third dimension is the "phantom" one. Again, it is important that we have the memory such that `t2[1].2` is a contiguous chunk, and `t2[2].2` follows it. I think we can accomplish this by doing the column major conversion not at the complete end of a key, but at the end of the array literals?


Does this make sense? I'm not sure I conveyed that super clearly

Reply:
I didn't realize it before, but this does imply that `array[2] tuple(array[3] real, array[4] int) x` actually _should_ be serialized differently than `tuple(array[2, 3] real, array[2, 4] int) x`. In the second case we'd want the column major behavior, but it doesn't really make sense in the first.

Reply:
great - given the behavoir you describe, this makes sense.

Reply:
question:  how will `write_array` serialize variable `array[2] tuple(array[3] real, array[4] int) x` and how will it serialize var `tuple(array[2, 3] real, array[2, 4] int) x`?

Reply:
`write_array`s only contract seems to be that it writes out parameters in the same order `constrained_param_names` does. And those two would have different orders of naming. I believe these are equivalent to the in-order traversal if you think of the type as a tree.

Currently, the first generates this as a CSV header:
```
x.1:1.1,x.1:1.2,x.1:1.3,x.1:2.1,x.1:2.2,x.1:2.3,x.1:2.4,x.2:1.1,x.2:1.2,x.2:1.3,x.2:2.1,x.2:2.2,x.2:2.3,x.2:2.4
```

The second will generate:
```
x:1.1.1,x:1.2.1,x:1.1.2,x:1.2.2,x:1.1.3,x:1.2.3,x:2.1.1,x:2.2.1,x:2.1.2,x:2.2.2,x:2.1.3,x:2.2.3,x:2.1.4,x:2.2.4
```

(as per the design doc, `:` indicates a tuple index, and `.` is still used for all other indices)

Unwinding all this in cmdstanpy/r will be...fun.

Reply:
fun indeed - but it only needs to be done once.